File: lib/go/policy/auto_ban.go
package policy

// auto_ban.go
// Per-IP violation tracking and automatic temporary ban injection.
//
// When an IP accumulates `threshold` violations within a sliding `window`, it is
// dynamically inserted into the RadixIP Engine blocklist with an "auto-banned"
// metadata value. The ban entry is expired by a background sweeper goroutine
// started on NewAutoBanTracker.

import (
	"net"
	"strconv"
	"sync"
	"time"

	config "github.com/Mwangi-Derrick/radixip/lib/go/config"
	engine "github.com/Mwangi-Derrick/radixip/lib/go/engine"
)

// BanEngine is the subset of the RadixIP Engine required by the AutoBanTracker.
type BanEngine interface {
	Insert(prefix *net.IPNet, meta engine.Metadata) error
	Remove(prefix *net.IPNet) *engine.Metadata
}

// AutoBanTracker records per-IP rate-limit violations and auto-bans repeat
// offenders by injecting them into the RadixIP blocklist tree.
type AutoBanTracker struct {
	mu          sync.Mutex
	violations  map[string][]time.Time // ipStr → sorted violation timestamps
	banned      map[string]time.Time   // ipStr → ban expiry time
	threshold   uint64
	window      time.Duration
	banDuration time.Duration
	engine      BanEngine
}

// NewAutoBanTracker creates and starts an AutoBanTracker backed by the given
// engine and config. The background sweeper goroutine is started automatically;
// call Stop() on shutdown.
func NewAutoBanTracker(cfg config.AutoBanConfig, eng BanEngine) *AutoBanTracker {
	t := &AutoBanTracker{
		violations:  make(map[string][]time.Time),
		banned:      make(map[string]time.Time),
		threshold:   cfg.ThresholdViolations,
		window:      time.Duration(cfg.WindowSeconds) * time.Second,
		banDuration: time.Duration(cfg.BanDurationSeconds) * time.Second,
		engine:      eng,
	}
	go t.sweeper()
	return t
}

// RecordViolation records a rate-limit violation for ipStr. If violations exceed
// the threshold within the sliding window, the IP is auto-banned and true is
// returned. Returns false otherwise.
func (a *AutoBanTracker) RecordViolation(ipStr string) bool {
	a.mu.Lock()
	defer a.mu.Unlock()

	now := time.Now()

	// Prune violations outside the window.
	windowStart := now.Add(-a.window)
	prev := a.violations[ipStr]
	filtered := prev[:0] // reuse slice memory
	for _, t := range prev {
		if t.After(windowStart) {
			filtered = append(filtered, t)
		}
	}
	filtered = append(filtered, now)
	a.violations[ipStr] = filtered

	if uint64(len(filtered)) >= a.threshold {
		a.ban(ipStr, now)
		return true
	}
	return false
}

// IsBanned reports whether ipStr is currently in a ban period managed by this
// tracker. Note: banned IPs are also in the engine blocklist, so the middleware
// blocklist check will catch them independently. This is a cheap local check.
func (a *AutoBanTracker) IsBanned(ipStr string) bool {
	a.mu.Lock()
	defer a.mu.Unlock()
	expiry, ok := a.banned[ipStr]
	return ok && time.Now().Before(expiry)
}

// ban inserts ipStr into the engine blocklist and records the expiry.
// Must be called with a.mu held.
func (a *AutoBanTracker) ban(ipStr string, now time.Time) {
	expiry := now.Add(a.banDuration)
	a.banned[ipStr] = expiry
	// Reset violation counter so repeated floods don't keep re-banning.
	delete(a.violations, ipStr)

	// Build the /32 (IPv4) or /128 (IPv6) host prefix for the engine.
	ip := net.ParseIP(ipStr)
	if ip == nil {
		return
	}
	var prefix *net.IPNet
	if ip.To4() != nil {
		_, prefix, _ = net.ParseCIDR(ipStr + "/32")
	} else {
		_, prefix, _ = net.ParseCIDR(ipStr + "/128")
	}
	if prefix == nil {
		return
	}
	_ = a.engine.Insert(prefix, engine.Metadata{
		Value: "auto-banned",
		Attributes: map[string]string{
			"reason":     "exceeded_threshold",
			"count":      strconv.FormatUint(a.threshold, 10),
			"window":     a.window.String(),
			"expires_at": expiry.UTC().Format(time.RFC3339),
		},
	})
}

// sweeper is a background goroutine that removes expired bans from the engine.
func (a *AutoBanTracker) sweeper() {
	tick := time.NewTicker(30 * time.Second)
	defer tick.Stop()
	for range tick.C {
		a.sweep()
	}
}

func (a *AutoBanTracker) sweep() {
	now := time.Now()
	a.mu.Lock()
	expired := make([]string, 0)
	for ipStr, expiry := range a.banned {
		// Treat bans whose expiry time is equal to or before now as expired.
		// Use !now.Before(expiry) instead of now.After(expiry) to include equality
		if !now.Before(expiry) {
			expired = append(expired, ipStr)
		}
	}
	for _, ipStr := range expired {
		delete(a.banned, ipStr)
	}
	a.mu.Unlock()

	// Remove from engine outside the lock to avoid deadlocking on engine ops.
	for _, ipStr := range expired {
		ip := net.ParseIP(ipStr)
		if ip == nil {
			continue
		}
		var prefix *net.IPNet
		if ip.To4() != nil {
			_, prefix, _ = net.ParseCIDR(ipStr + "/32")
		} else {
			_, prefix, _ = net.ParseCIDR(ipStr + "/128")
		}
		if prefix != nil {
			a.engine.Remove(prefix)
		}
	}
}

File: lib/rust/policy/src/auto_ban.rs
//! Per-IP violation tracking and automatic temporary ban injection.
//!
//! When an IP accumulates `threshold_violations` rate-limit violations within
//! a sliding `window_seconds` window, it is automatically inserted into the
//! RadixIP blocklist engine as a /32 (IPv4) or /128 (IPv6) host route with
//! value `"auto-banned"`.  A background thread sweeps expired bans and
//! removes them from the engine.
//!
//! ## Concurrency model
//!
//! Violations are tracked in a `DashMap<IpAddr, Mutex<Vec<Instant>>>`.
//! Each IP has its own per-entry lock, so concurrent requests from *different*
//! IPs never contend with each other. The ban table is a plain
//! `DashMap<IpAddr, Instant>` where reads (the hot path) only hold a short
//! shard read-lock for a few nanoseconds.
//!
//! This is safe to use from both async Tokio contexts and sync FFI threads
//! (Python/Node) because no blocking I/O is performed under any lock.

use std::net::IpAddr;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use dashmap::DashMap;
use ipnetwork::IpNetwork;
use radixip::RadixEngine;
use radixip_config::AutoBanConfig;

// AutoBanTracker

/// Tracks per-IP violations and injects auto-bans into a `RadixEngine`.
///
/// Cheap to clone — all state lives behind `Arc`.
#[derive(Clone)]
pub struct AutoBanTracker {
    /// Per-IP violation timestamps (each IP holds its own short-lived lock).
    violations: Arc<DashMap<IpAddr, Mutex<Vec<Instant>>>>,
    /// Per-IP ban expiry. Reads only need a shard read-lock (very fast).
    banned: Arc<DashMap<IpAddr, Instant>>,
    threshold: u64,
    window: Duration,
    ban_duration: Duration,
    engine: Arc<Box<dyn RadixEngine>>,
}

impl AutoBanTracker {
    /// Create a new tracker and start the background expiry sweeper.
    pub fn new(cfg: &AutoBanConfig, engine: Arc<Box<dyn RadixEngine>>) -> Self {
        let violations: Arc<DashMap<IpAddr, Mutex<Vec<Instant>>>> =
            Arc::new(DashMap::new());
        let banned: Arc<DashMap<IpAddr, Instant>> = Arc::new(DashMap::new());

        let tracker = Self {
            violations: Arc::clone(&violations),
            banned: Arc::clone(&banned),
            threshold: cfg.threshold_violations,
            window: Duration::from_secs(cfg.window_seconds),
            ban_duration: Duration::from_secs(cfg.ban_duration_seconds),
            engine: Arc::clone(&engine),
        };

        // Spawn background sweeper on a plain OS thread — works in both Tokio
        // and non-Tokio environments (Python/Node FFI).
        {
            let banned_clone = Arc::clone(&banned);
            let engine_clone = Arc::clone(&engine);
            std::thread::spawn(move || {
                sweeper(banned_clone, engine_clone);
            });
        }

        tracker
    }

    /// Record a rate-limit violation for `ip`.
    ///
    /// Returns `true` if the IP was auto-banned as a result of this call.
    /// Only holds the per-IP entry lock for the duration of the prune+count
    /// operation — never a global lock.
    pub fn record_violation(&self, ip: IpAddr) -> bool {
        let now = Instant::now();
        let cutoff = now.checked_sub(self.window).unwrap_or(now);

        // Get-or-insert the per-IP bucket (DashMap shard lock only).
        let entry = self
            .violations
            .entry(ip)
            .or_insert_with(|| Mutex::new(Vec::new()));

        {
            let mut timestamps = entry.value().lock().unwrap();

            // Prune old entries outside the sliding window.
            timestamps.retain(|&t| t >= cutoff);
            timestamps.push(now);

            if (timestamps.len() as u64) < self.threshold {
                return false; // lock released here
            }

            // Threshold reached — reset counter so a continuous flood doesn't
            // keep re-triggering the ban on every subsequent violation.
            timestamps.clear();
        } // per-IP lock released here

        // Record the ban and insert into the engine (no locks held).
        let expiry = now + self.ban_duration;
        self.banned.insert(ip, expiry);
        drop(entry); // release DashMap entry ref before engine call
        self.insert_ban(ip);
        true
    }

    /// Returns `true` if `ip` is within an active auto-ban period.
    ///
    /// Fast path: only acquires a DashMap shard read-lock for nanoseconds.
    pub fn is_banned(&self, ip: IpAddr) -> bool {
        self.banned
            .get(&ip)
            .map(|expiry| Instant::now() < *expiry)
            .unwrap_or(false)
    }

    fn insert_ban(&self, ip: IpAddr) {
        let prefix = host_prefix(ip);
        let meta = radixip::types::Metadata::new("auto-banned")
            .with_attribute("reason", "exceeded_threshold");
        let _ = self.engine.insert(prefix, meta);
    }
}

// Background sweeper

/// Runs forever on a background OS thread. Every 30 s it removes expired
/// bans from both the in-memory map and the RadixIP engine.
fn sweeper(banned: Arc<DashMap<IpAddr, Instant>>, engine: Arc<Box<dyn RadixEngine>>) {
    loop {
        std::thread::sleep(Duration::from_secs(30));

        let now = Instant::now();

        // Collect expired IPs (brief iteration scan over DashMap shards).
        let expired: Vec<IpAddr> = banned
            .iter()
            .filter(|r| now >= *r.value())
            .map(|r| *r.key())
            .collect();

        // Remove from ban map and engine (no locks held during engine calls).
        for ip in &expired {
            banned.remove(ip);
        }
        for ip in expired {
            let prefix = host_prefix(ip);
            engine.remove(&prefix);
        }
    }
}

// Helpers

/// Build a /32 (IPv4) or /128 (IPv6) host network for engine insertion.
fn host_prefix(ip: IpAddr) -> IpNetwork {
    match ip {
        IpAddr::V4(v4) => {
            IpNetwork::V4(ipnetwork::Ipv4Network::new(v4, 32).expect("32-bit mask always valid"))
        }
        IpAddr::V6(v6) => {
            IpNetwork::V6(ipnetwork::Ipv6Network::new(v6, 128).expect("128-bit mask always valid"))
        }
    }
}

